### PR TITLE
extend dlist to detect unlinked nodes and avoid double-remove bugs

### PR DESCRIPTION
**Pull request by pabigot**
This PR marks `sys_dnode_t` instances that are not in a list so that code does not depend on overloaded external state to assess this condition before potentially violating dlist API requirements.

A second commit uses that API to avoid a double-remove in timeout that could result in a corrupted list of timers, as an alternative solution to a commit in #12204 (for full context see [this discussion](https://github.com/zephyrproject-rtos/zephyr/pull/12204/commits/0923b5dff9cb3da2f766655d755bdebcf176fd6b#r243822029) if that doesn't disappear when the commit goes away.)

A third commit works around an unknown double-remove bug in the kernel poll infrastructure.  Why the events are being removed multiple times should probably be investigated by a kernel maintainer.

### CODE REVIEW

File: kernel/timeout.c
Comment (pabigot) [inline kernel/timeout.c]: @andyross Please review.  In particular, I was surprised that it's possible for a timeout structure to be linked while it's marked inactive.  That could happen legitimately if the timeout is linked into a different list (e.g. of unused timeouts), but I'm not seeing how that could be.

If the check for `dticks` is removed then several sanity checks fail, including `samples/cpp_synchronization/test`.

**Review by andyross — COMMENTED** [review]
Generally in favor, though it would be good to get a feel for how many list operations are being affected and what the performance and code size impact (sys_dlist_remove is generally inlined) is likely to be.

But yeah, we need to fix and not work around that spot in timer that's failing.

File: kernel/poll.c
Comment (pabigot) [reply kernel/poll.c]: `sys_dlist_remove` already documents that the node must be part of the list, and in most situations that's true (and known at the call-site).  So I don't think there's reason to add a check in that function.

File: kernel/timeout.c
Comment (pabigot) [reply kernel/timeout.c]: [pabigot/zephyr:pr/20181230b](https://github.com/pabigot/zephyr/tree/pr/20181230b) is this branch rebased on #12204 after removing your workaround.  With the following change it fails `cpp_synchronization`.  So the problem is independent of whether #12204 is merged.
```
diff --git a/kernel/timeout.c b/kernel/timeout.c
index ffe6e78039..6ea334f461 100644
--- a/kernel/timeout.c
+++ b/kernel/timeout.c
@@ -96,8 +96,8 @@ int _abort_timeout(struct _timeout *to)
        int ret = _INACTIVE;
 
        LOCKED(&timeout_lock) {
-               if (sys_dnode_is_linked(&to->node)
-                   && (to->dticks != _INACTIVE)) {
+               if (sys_dnode_is_linked(&to->node)) {
+                       printk("remove %p, %d\n", &to->node, to->dticks);
                        remove_timeout(to);
                        ret = 0;
                }
```

Comment (pabigot) [reply kernel/timeout.c]: I found a "hidden" `_init_timeout` function that needed to be updated so a stack-allocated `k_timer` in the C++ sample didn't start with bogus pointers.  So the secondary check is no longer present, and it'd probably be possible to eliminate `_INACTIVE` as a flag value to detect disabled timeout structures.

**Comment by pabigot** [comment]
Rebased on top of merged #12204, added commit to remove use of `_INACTIVE` as a flag value to detect timeouts that are not active.

File: kernel/sched.c
Comment (andyross) [inline kernel/sched.c]: Nitpick on behalf of @ceolin : MISRA rules disallow implicit conversion to bool

But I think this is deeper.  The old code would store the "was linked" state implicit in the timeout, but that's no longer ture, right?  You don't know that a zero return here means "inactive", I think?   Might need to be reworked to do a is_linked call first and then use that as the predicate for the if()?

Actually a quick grep shows that this one of two places actually using the return value of _abort_(thread_?)timeout().  Maybe the simpler thing to do would be to just remove that return value and/or swap it to a boolean that indicates "was active".

File: kernel/include/timeout_q.h
Comment (andyross) [inline kernel/include/timeout_q.h]: Not particularly relevant, but this was always a bad API (a true return indicates a false state) that never got fixed up during all the refactoring.  I wouldn't cry to see this patch change it to z_timeout_active() or whatever.

File: kernel/timeout.c
Comment (andyross) [inline kernel/timeout.c]: As mentioned elsewhere, probably best to just eliminate the return value here or make it a boolean to indicate "was active", as that's all existing callers care about.

**Review by andyross — CHANGES_REQUESTED** [review]
One spot I think is wrong, a few other suggestions.

And I do still think we should pull up some size comparisons on stuff like tests/benchmarks/footprint to see how much this affects code size.  I'd guess that this is a worthwhile tradeoff too, but it'd be nice to prove that.

File: kernel/sched.c
Comment (pabigot) [reply kernel/sched.c]: The old code returned zero on successful abort, and `_INACTIVE` on failure.  Returning `-EINVAL` is more consistent with indicating that the abort failed.  I really think this is doing the right thing.

I'll change the test to detect a negative result, though, instead of implicit bool although there's an awful lot of implicit bool in Zephyr.

File: kernel/include/timeout_q.h
Comment (pabigot) [reply kernel/include/timeout_q.h]: This was partly because of the legacy of `_EXPIRED`: timeouts could be in three states, not two.  This function tests that the state is exactly what `_INACTIVE` indicated.

A true result indicates that the predicate identified by the function name is true.  Why is that a bad API?  We use `sys_dlist_is_empty()`, not `sys_dlist_is_nonempty()`.

The whole timeout API uses prefix underscore.  While I'd rather that wasn't the case (violates C reserved identifiers for the non-static functions) I wouldn't want to change just one function, nor change all of them, in this PR.

File: kernel/timeout.c
Comment (pabigot) [reply kernel/timeout.c]: The return indicates the attempt to abort the timeout failed because the timeout wasn't active (valid for abort).  I believe this is the preferred way to indicate the result of the operation.

**Comment by andrewboie** [comment]
Good stuff here, although I share Andy's concerns about performance impact since dlists tend to be pretty hot code paths, perhaps consider ifdefing the extra checking around CONFIG_ASSERT or something like that?

**Comment by pabigot** [comment]
There are two concerns: impact on code size from the added null assignments, and impact on performance of the null assignments (and, rarely, checks; most of the time the linked state does not need to be checked).

I can run comparisons for code size, if somebody tells me what platform(s) and what test application(s) to use.  What are the criteria for acceptable/unacceptable impact of the change?

Performance is a different issue.  First, I'd need instruction on how that would be evaluated in a meaningful way.

Second, the changes here have two effects: bugs where lists would be corrupted by double removes are converted from undetected corruption to faults at point-of-use; and a way to reliably identify nodes that are not in a list, as is needed with timers.  The first is a quality issue, and we could certainly turn it off in some cases.  The second is a functional capability, which would break if it were optional.

Note that the Linux dlist implementation invalidates the pointers when nodes are removed.  They seem to believe any performance impact is negligible relative to the value of detecting failures.

What do you want me to do?

**Comment by andrewboie** [comment]
> Note that the Linux dlist implementation invalidates the pointers when nodes are removed. They seem to believe any performance impact is negligible relative to the value of detecting failures.

FWIW, Linux isn't targeting microcontrollers with memory on the order of 64K (or less)

>  way to reliably identify nodes that are not in a list, as is needed with timers.

Fair enough!

**Review by tyrantlizard77 — APPROVED** [review]

**Review by andyross — APPROVED** [review]
Removing -1, as I'm convinced the behavior looks correct. I'd still prefer to see _abort_timeout()'s signature matched to the needs to the needs of its actual callers instead of trying to return an error code (it's an internal function, not an API for users outside of the kernel -- hell, even within the kernel it's used only at the lowest level), but that can be fixed whenever.

**Comment by pabigot** [comment]
@andrewboie @andyross Thanks for the approvals.  Please take the following into account before merging:

The intermittent failure of mps2_an385 on `tests/kernel/mem_slab/mslab_threadsafe/kernel.memory_slabs` is due to MPU or BUS errors derefencing invalidated node link pointers during a call `_priq_dumb_remove`.

By instrumenting the code I can tell that `thread->base.qnode_dlist` (which is actually used as `dnode`) has non-null `next` and `prev` pointers prior to invoking `sys_dlist_remove` but at least one null `next` or `prev` pointer within `sys_dlist_remove`.  (Neither is a pointer-to-self, either, which would produce this behavior.)

This appears to be a bug identified by this PR, not one introduced by this PR, although the cause is unclear.  Unlike the poll bug no workaround seems to be effective.  I don't know how to debug this further.

File: kernel/include/timeout_q.h
Comment (andyross) [reply kernel/include/timeout_q.h]: > A true result indicates that the predicate identified by the function name is true

I don't think that not incorrect, exactly, but it's not innapropriate to argue against the ellision of unommitted negatives either.

No seriously, double negatives in API choices are bad, especially in predicates.  The base adjective for that state is "active".  True should mean "active", false should mean "inactive".

**Review by andyross — CHANGES_REQUESTED** [review]
Oh, wait.  Sorry, I missed that warning earlier.  Yeah, we for sure need to get any introduced failures addressed before this merges.   If you want me to look at it, can you put together a reproducer and file it as a bug?

**Comment by pabigot** [comment]
Your approvals came in while I was doing final edits on the warning; glad you caught it.

https://github.com/pabigot/zephyr/commits/diag/20181230a has a diagnostic commit on top of this PR.  The commit message describes how to reproduce the problem.  (The `10563` error code indicates the failing call came from line 563 of the instrumented `sched.c` under branch condition 10000.)

Thanks for looking at this.

**Comment by pabigot** [comment]
@andyross Thanks for looking at this.  Your description provided some helpful clues.

I was able to replicate the observation of `_current` having null links in the context of the timer ISR.  In my case it occurred because the main thread had exited (removing itself from all run queues) but was still recorded as the value of `_current`.  The tick ISR was still active and when it fired tried moved `_current` to the back of the queue, which caused havoc since it wasn't in a queue at that point.  There might be a better fix, assuming this situation can ever occur in an unhacked kernel; nonetheless I resolved it by changing `sliceable` to ignore unrunnable threads.

After that red herring was properly tinned I finally found the real problem: mem_slab retains a queue of threads that want to allocate from it that should be scheduled on memory release, but the operation to pull threads from a wait queue didn't handle the case where the queue was empty, at which point it queued garbage.  That's also been fixed.

Finally, I corrected the name of the link field in `_thread_base` so we no longer appear to be appending lists to lists.

**Comment by pabigot** [comment]
My last patch set solved the problem, by breaking timeslicing.  The proper explanation and solution is in this commit:

>    kernel: sched: inhibit timeslice from unready thread
>    
>    If timeslicing is enabled `z_clock_announce` will attempt to swap the
    current thread for another.  The implementation of `z_time_slice`
    assumes that `_current` is always in the queue, and so moves it, which
    relies on valid links in the dnode structure.
>    
>    When the current thread is in the process of pending itself the thread
    may have already been removed from the queue at the point the tick ISR
    runs.  This violates the requirement in `_move_thread_to_end_of_prio_q`
    that the thread already be in the queue.
>    
>    If the current thread is not queued treat it as unsliceable: it's in the
    process of being swapped anyway.

I'm leaving this one in place for now:

>    kernel: sched: avoid potential null pointer dereference in wait queue
>    
>    The existing code has a too-subtle reliance on the sys_dnode_t instance
    being located at the start of its container: this is the only situation
    where `CONTAINER_OF(NULL, struct, tag)` will be guaranteed to produce a
    null pointer.
>    
>    When this required layout is violated, e.g. during diagnostics, the
    system fails horribly: for example an empty wait queue will result in an
    attempt to operate on a thread that does not exist.
>    
>    Update all _priq_foo_best implementations to explicitly check for a null
    pointer rather than relying on the dnode being first in the thread
    structure.

because the original code was way too subtle, wasting time for both @andyross and me.  However, the checks cost 12 bytes to add under SDK 0.9.5, and 16 bytes under gcc 8.2.1, so you can tell me to remove it.

File: kernel/sched.c
Comment (andyross) [inline kernel/sched.c]: The fix is correct, but can you clarity the explanation?  It was a bug, call it a bug, e.g:

    kernel/sched: Fix empty list detection

    CONTAINER_OF() on a NULL pointer returns some offset around NULL and
    not another NULL pointer.  We have to check for that ourselves.

    This only worked because the dnode happened to be at the start of the struct.

File: kernel/sched.c
Comment (andyross) [inline kernel/sched.c]: This sounds wrong.  A thread "in the process of pending itself" surely shouldn't be interrupted by a timer interrupt.  The time period between the list removal and the _Swap() should be under a lock, and I'm pretty sure I chased that last night and convinced myself it was true.

I buy that this works to work around the bug, but I don't think it's a root cause.  We need to find out why we're in this state (i.e. point to a line of code and say "this is where it can be interrupted while unqueued") and then make sure that's locked or fix the design bug that prevents it.

**Review by andyross — CHANGES_REQUESTED** [review]
Nice catch on the CONTAINER_OF() goof (though, heh, the commit message complaining about subtle code was too subtle for my tastes).

I don't think we're quite there yet on the crash bug, though.   This would make sense to me as a workaround for an existing bug, but not just to merge new code.  I'll take a closer look later too.

File: kernel/sched.c
Comment (pabigot) [reply kernel/sched.c]: What appears to be happening is one of the worker threads in the test completes and exits, which causes k_thread_abort() to be invoked on it while it is still `_current`.  This does occur under lock, terminated by a call to __swap().

__swap() twiddles some state, then releases the IRQ lock and relies on a triggered PendSV handler to complete the context switch and update `_current`.

Before that can happen, a higher priority timer ISR fires which initiates a timeslice.  The aborted thread is still in `_current`, which is no longer in any queue, and if the slice is taken a double-remove is issued.  Prior to this PR that didn't cause any visible harm; now it does.

The proposed patch works because it blocks taking the slice from an unqueued `_current`.  From what I can tell this is a pre-existing bug, revealed but not introduced by this PR, so the workaround should be acceptable within the scope of this PR.

I've changed the commit message of the other commit as requested.

**Comment by andyross** [comment]
Quick discussion with @pabigot  on skype convinced me I see the arc of what's happening here:

Background: _Swap() on ARM is nonatomic.  It works by triggering a PendSV exception (which is a flagged exception that's really an "interrupt", you set the bit and then the trap will be taken at the next point where the interrupt state allows it) which effects the swap by setting all the appropriate register and stack state and then returning into the next thread.

BUT: PendSV is actually low priority and can itself be preempted (at the beginning of its execution before the hadnler locks interrupts on its own).  So a hardware interrupt can arrive **before** the swap actually happens, in a state where the old _current has been removed from the list but the new one has not been assigned into the _current pointer.

And boom: that's all that's needed to reproduce the behavior we're seeing.  If it's a timer interrupt and the slice timer expires (and in this particular test interrupts arrive at 1kHz and the slice timeout is set at a single 1ms tick!) it will see the unexpectedly unqueued _current and explode on the now-more-robust dlist code that isn't expecting that.

Now I just need to think about a simple enough workaround.  Honestly I'd be really tempted to change the priority of PendSV to the system maximum priority as this is now the **third** bug resulting from that situation.  But that really needs to be done by ARM people; there may be good reasons for the priority being what it is.

Beyond that we can certainly treat the timeslice issue with a special case ARM test to see if a PendSV is on the stack below us, I guess.  But even that's a little fiddly -- we can set a global variable out of _Swap() I guess and reset it in the assembly, maybe...  Let me think a bit, or tell me what you come up with.

**Comment by andyross** [comment]
OK, took almost as much puzzling to figure out how to push to someone else's PR as it did to implement the workaround.  I think I got it right.

This removes the original timeslice workaround in favor of a targetted one that addresses the root cause, and only on affected systems.  Includes a CONFIG_SWAP_NONATOMIC kconfig to enable it on ARM only.  And it also refactors @mike-scott 's fix from the scheduler rewrite so that it's now enabled on ARM only (at the time, we didn't understand the root cause).

The hacked/whiteboxed failure case I had in that mslab test runs indefinitely for me now, and I got it through 65ish runs under sanitycheck without a failure, so I'm reasonably certain this really was the root cause and that this is squashed now.

**Review by andyross — DISMISSED** [review]

**Comment by andyross** [comment]
Last minute change to make the flag a boolean broke a timeslicing test -- it's not enough to know you just pended "something", because you might run AFTER the context switch too, check that the pended thread still matches _current.  Now it works.

**Comment by pabigot** [comment]
Github doesn't allow me to review my own patch, so I'm marking this DNM explicitly and hoping that's noticed in time, my alternative being to close the PR.

A workaround that I provided for a test failure has been replaced by a set of three commits that fix the underlying cause of the failure, which was an unreported infrastructure bug.  The proposed fix over three commits is a reasonable solution for the root cause, though if I could review it I would make some suggestions.

The showstopper is that if merged through this PR these changes would be baselined with approvals from the person who added them and somebody who reviewed the PR before they were added.

This would be a clear violation of the spirit if not letter of the two-sign-off rule Zephyr is using.

@andyross opened #12342 for this (but didn't link to it here or in the commit messages), and I've commented there.  This PR should be held until the underlying bug blocking it is fixed following the proper process.

**Comment by andyross** [comment]
Um... OK.  That's a valid point.  I'll dismiss my review.  I genuinely thought this was how you wanted me to do this.  Would you prefer I submit your patches out of my own PR?  Help me out here.  I swear, I like your dlist changes; I want them in the tree, and I'm trying to get them merged.

But #12342 isn't a bug.  That's the way _Swap() works on ARM and always has.  I'd personally like to change it, but I don't own the ARM architecture and am not the relevant expert to make the call.

**Comment by andyross** [comment]
> @andyross opened #12342 for this (but didn't link to it here or in the commit messages),

Just FWIW: I actually did.  It looks like I messed it up somehow and the link in the description isn't actually a link.  Profuse apologies, seriously.  I promise I'm not trying to exclude you from anything.

**Comment by carlescufi** [comment]
@pabigot please rebase

**Comment by carlescufi** [comment]
@andyross if you are happy with this please approve so we can move this forward

**Comment by pabigot** [comment]
@andrewboie @andyross in the churn of this PR the following additional commit got added after the original review process.  If this is OK I'd like to leave it, if not I can pull it out and resubmit it in another PR that's blocked by the dlist features.

### kernel: timeout: detect inactive timeouts using dnode linked state

Whether a timeout is linked into the timeout queue can be determined
from the corresponding sys_dnode_t linked state.  This removes the need
to use a special flag value in dticks to determine that the timeout is
inactive.

Update _abort_timeout to return an error code, rather than the flag
value, when the timeout to be aborted was not active.

Remove the _INACTIVE flag value, and replace its external uses with an
internal API function that checks whether a timeout is inactive.

**Comment by pabigot** [comment]
@andrewboie @andyross @carlescufi Actually, hold on: to make things easier I'm going to close this one, and submit the one that has the combined dlist and timer updates for #12332 together.

**Comment by pabigot** [comment]
Reopening since Andy doesn't like the part of #12485 that motivated putting it in a separate PR.  So this is back to being the original content, plus the removal of the magic `_INACTIVE` flag value.

**Comment by pabigot** [comment]
@andyross Reverted the change of the structure tag per https://github.com/zephyrproject-rtos/zephyr/pull/12485#discussion_r248018705

**Comment by pabigot** [comment]
@andyross this is back to the original submitted content, plus the removal of the inactive magic value and restoring the `qnode_dlist` spelling.  Is it OK now?

**Comment by carlescufi** [comment]
@andyross can you please review the last commit in this series?

**Review by andyross — APPROVED** [review]

**Comment by carlescufi** [comment]
Thanks @andyross

**Pull request merged**
